samples: sockets: socketpair: sample application and docs

cfriedt opened the pull request:
Support for the socketpair(2) system call was recently added for 2.3.0.

This change adds a sample application that demonstrates how it can be used.

Fixes #25527

cfriedt commented:
Was maybe being a bit hopeful for `std::thread`, although it should be relatively simple to added that support. I think it's more of a newlib issue than anything. Also `std::chrono` ... maybe less simple.

jukkar commented on samples/net/sockets/socketpair/src/socketpair_example.cpp:
```diff
@@ -0,0 +1,184 @@
+/*
```
Why C++? At least I would prefer a plain C program as a sample.

pfalcon replied:
> Why C++? At least I would prefer a plain C program as a sample.

I'm kinda surprised too. But without people bringing up new ideas, we won't develop as wide and far as we could. We could treat this sample as a cross between showcases socketpair and C++ functionality, and we clearly need more C++ samples. And if this sample really builds under both a POSIX system and Zephyr, IMHO that's rather positive development overall.

cfriedt replied:
I find C++ just a nice shorthand for C requiring fewer lines of code - especially when looking at the various container types. It's more expressive.
I could potentially convert it to C, but at the same time.. if it Just Works™.
It's just a sample program but I too thought it would be good to start demonstrating C++ code running on Zephyr too from a code diversity perspective. There are a lot of devs who *want* to write C++ code on their microcontrollers, so maybe it would be good advertisement ¯\_(ツ)_/¯.

jukkar replied:
I would like to see a sample implemented using C like the other network samples. That way anyone writing a C program can use this sample right away. Having a C++ sample here is fine too, we can have two implementations.

pfalcon replied:
> we can have two implementations.

@cfriedt, So I hope you'll keep C++ variant, as I mentioned, IMHO it's useful.

cfriedt replied:
I have a separate PR now for multithreading in C++, so adding some samples for C++ application support is a no-brainer. I'll rewrite this one in C.

cfriedt commented:
Hehe... so looks like I'm implementing nanosleep. Probably better to do that in a separate PR ;-)

cfriedt commented:
I'd like to leave this as a draft until #25572 is merged.

pfalcon commented:
> I'd like to leave this as a draft until #25559 and #25572 are merged.

To set expectations right, this may be as well after the 2.3 release (and surely, not e who decides).

cfriedt commented:
> > I'd like to leave this as a draft until #25559 and #25572 are merged.
> 
> To set expectations right, this may be as well after the 2.3 release (and surely, not e who decides).
Of course ;-)

jukkar reviewed (approved):
Thanks, looks good!

pfalcon commented:
From https://github.com/zephyrproject-rtos/zephyr/pull/25528#issuecomment-632666714

> @pfalcon saw #24730

And I didn't saw this and following lines. Because they were edited into an already posted comment, and the original content of that comment is what I got in email and considered it read. I mentioned this issue lately elsewhere. Please avoid editing comments for adding new information, post it as a new comment instead. It's ok to post a bunch of comments in row (if it's needed). (And I for one edit my comments, but only to fix typos/ambiguous wording).

cfriedt commented:
It looks like there was a copy-paste error made several years ago that is breaking this PR.

https://github.com/zephyrproject-rtos/zephyr/commit/c8aa6570c14dfc64ccfafd2c73a924bebc227a56

```
typedef struct pthread_attr_t {
...
} pthread_attr_t;
```

pfalcon commented:
@cfriedt: I don't see a copy-paste error there. It's quite a natural way to use the same name for both `struct` and `typedef`, but if you hint that it plays bad with `#define` which is supposed to target just one of those names, then yeah, could be.

cfriedt commented:
@pfalcon - it's specific to nrf52bsim I think. I didn't see the entirety of the error message... but posix types are defined twice.
```
-include /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/arch/posix/include/posix_cheats.h
```

cfriedt committed:
samples: sockets: socketpair: sample application and docs

Support for the socketpair(2) system call was recently
added for 2.3.0 .

This change adds a sample application that demonstrates
how it can be used.

Fixes #25527

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

pfalcon commented:
nrf52bsim is another instance of CONFIG_ARCH_POSIX, aka native_posix issue-alike.

Anyway, feel free to propose a fix. As usual, submitting it as a separate PR raises chances for it to be processed sooner rather than later (but I assume, @carlescufi is busy with high-priority fixes by now).

cfriedt commented:
@carlescufi  - sample application and docs would be good to accompany socketpair for 2.3.0. I'm sure it's too late now, but just thought I would throw it out there.

pfalcon commented:
> @carlescufi - sample application and docs would be good to accompany socketpair for 2.3.0. I'm sure it's too late now, but just thought I would throw it out there.

Given that socketpair() is just introduced for 2.3, we hardly can make it worse for anyone with further changes, only better. @carlescufi, adding 2.3 milestone to help you not forget about this case ;-).

carlescufi commented:
I am merging this because it's a sample that demonstrates a new feature that was introduced in 2.3.0.
